contracts/AMMStrategy.sol

```diff
@@ -0,0 +1,181 @@
+            ammMapping[_ammAddrs[i]] = true;
+            emit SetAMM(_ammAddrs[i], true);
+        }
+    }
+
+    /************************************************************
+     *                 Internal function modifier                *
+     *************************************************************/
+    modifier onlyEntryPoint() {
+        require(msg.sender == entryPoint, "only entry point");
+        _;
+    }
+
+    /************************************************************
+     *           Management functions for Owner               *
+     *************************************************************/
+    /// @inheritdoc IAMMStrategy
+    function setAMMs(address[] calldata _ammAddrs, bool[] calldata _enables) external override onlyOwner {
+        for (uint256 i = 0; i < _ammAddrs.length; ++i) {
+            ammMapping[_ammAddrs[i]] = _enables[i];
+            emit SetAMM(_ammAddrs[i], _enables[i]);
+        }
+    }
+
+    /// @inheritdoc IAMMStrategy
+    function approveTokens(
+        address[] calldata tokens,
+        address[] calldata spenders,
+        bool[] calldata usePermit2InSpenders,
+        uint256 amount
+    ) external override onlyOwner {
+        require(spenders.length == usePermit2InSpenders.length, "length of spenders not match");
+        for (uint256 i = 0; i < tokens.length; ++i) {
+            for (uint256 j = 0; j < spenders.length; ++j) {
+                if (usePermit2InSpenders[j]) {
+                    // The UniversalRouter of Uniswap uses Permit2 to remove the need for token approvals being provided directly to the UniversalRouter.
+                    _permit2Approve(tokens[i], spenders[j], amount);
+                } else {
+                    IERC20(tokens[i]).safeApprove(spenders[j], amount);
+                }
+            }
+        }
+    }
+
+    /// @inheritdoc IAMMStrategy
+    function withdrawLegacyTokensTo(address[] calldata tokens, address receiver) external override onlyOwner {
+        for (uint256 i = 0; i < tokens.length; ++i) {
+            uint256 selfBalance = Asset.getBalance(tokens[i], address(this));
+            if (selfBalance > 0) {
+                Asset.transferTo(tokens[i], payable(receiver), selfBalance);
+            }
+        }
+    }
+
+    /************************************************************
+     *                   External functions                      *
+     *************************************************************/
+    /// @inheritdoc IStrategy
+    function executeStrategy(
+        address inputToken,
+        address outputToken,
+        uint256 inputAmount,
+        bytes calldata data
+    ) external payable override onlyEntryPoint {
+        Operation[] memory ops = abi.decode(data, (Operation[]));
+        require(ops.length > 0, "empty operations");
+        require(inputAmount > 0, "empty inputAmount");
+        uint256 balanceBefore = Asset.getBalance(outputToken, entryPoint);
```

Comment:
Since all of the balance will be transferred before return, there's no need to calculate before/after balance diff?

Reply:
It is need just when `emit Swapped(inputToken, inputAmount, outputToken, balanceAfter - balanceBefore);`
If we don't emit the Swap event, It's ok.

Reply:
The `balanceAfter - balanceBefore` here is equal to `erc20.balanceOf(address(this))`.

Reply:
After discussion, remove `Swapped` event here and leave it to `GenericSwap`.